x-pack/plugins/reporting/server/lib/tasks/execute_report.ts

```diff
@@ -287,6 +292,12 @@ export class ExecuteReportTask implements ReportingTask {
     return report;
   }
 
+  // Generic is used to let TS infer the return type at call site.
+  private async throwIfKibanaShutsDown<T>(): Promise<T> {
+    await this.reporting.getKibanaShutdown$().toPromise();
```

Comment:
Given that the system is shutting down, I am not sure if async functions can be trusted to work 100% of the time. I think it would help if this was rearranged to have `throwIfKibanaShutsdown` be a synchronous function.

We could have a field in the class such as `.inShutdown` , which is updated by a subscriber of `this.reporting.getKibanaShutdown$()`

```typescript
constructor() {
  this.inShutdown = false;
  this.reporting.getKibanaShutdown$().subscribe(() => {
    this.inShutdown = true;
  });
}

private throwIfKibanaShutsDown<T>(): T {
  if (this.inShutdown) {
    throw new KibanaShuttingDownError();
  }
}
```

Reply:
Hm, that is an interesting point. I've not observed an issue coming from having `async` in this case. My understanding from what we have in the current version is that under-the-hood something similar is happening to your refactored version:

We are `await`ing the completion of an observable that will syncly call the subscriber once it is done (resolving the promise). Resolving a function runs the remainder of the code `sync`. Perhaps this demonstrates what I mean:

```js
const rx = require('rxjs');
const f = () => {
	return new Promise((resolve) => {
		console.log('first');
		resolve();
	});
}	

const fsync = () => {
	rx.from(f()).toPromise();
	console.log('second');
}

fsync();

// Always logs
// => first
// => second
```

I think the biggest risk is the async network request for updating the job, but there is no of getting around that.

Out of curiosity, do you know how we might test `async` vs `sync` implementations in this scenario?

Reply:
>  I've not observed an issue coming from having async in this case.

It seems that when we use a signal to "stop plugins gracefully" there is more that can be done than I realized. I should have understood it based on the fact the reports are getting the error code saved.

Let's keep the code as the way you have it, since it is a bit cleaner than my suggestion.

> Out of curiosity, do you know how we might test async vs sync implementations in this scenario?

Not really - not without getting much deeper than I care to at the moment :stuck_out_tongue_winking_eye: